### ISSUE TITLE
Very low performance for IMap.set

### ISSUE THREAD

**Issue by wsw2016** [hazelcast/hazelcast-cpp-client#465]
If using IMap.set with values larger than 1 MByte the most limiting factor for performance shall be the network bandwidth. On Ubuntu 18.04 in VirtualBox and i5-6300U I first checked the achievable TCP performance on localhost with "qperf localhost tcp_bw tcp_lat":
bw  =  4.26 GB/sec and 
latency  =  56 us.

A good middleware would be able to transfer big key-value pairs with 95% of the TCP bandwidth (e.g. a DDS implementation from RTI).

With the C++-client using the simple IMap<string, vector<uint8_t> >.set(key, value) method I got reproduceable 1.4 MBytes/Sec for values from 1 to 16 MBytes size (s. attachment).

First, I was curious, if Hazelcast as whole has so low performance and checked the same test for the java client and the same instance of Hazelcast-Server. The throughput was a little bit instable, but in the range of 200 MBytes/Sec. It is still only 5% of the available network bandwidth, but factor 140 higher than C++ implementation.

Interestingly the IMap.get in C++ has higher performance of around 90 MBytes/Sec.

In the tests around IMap in C++ I have not found any one checking performance especially for larger kv-pairs, perhaps it is a good idea to put one.

The Cpp-library was built with
cmake .. -DHZ_LIB_TYPE=STATIC -DHZ_BIT=64 -DCMAKE_BUILD_TYPE=Release

In the attachment:
- main.cpp with the c++ test
- Client.java with java test (from code-samples)
- cppIMapPerformance.txt with columns
	messageSize, setDurationInMicroSeconds, setThrouput, getDurationInMicroseconds, getThroughput
- javaIMapPerformance.txt the same for java, but with Milliseconds resolution
- JavaIMapErrOutput.txt output from error-stream during the test run.

[cppIMapPerformance.txt](https://github.com/hazelcast/hazelcast-cpp-client/files/2533226/cppIMapPerformance.txt)
[javaIMapErrOutput.txt](https://github.com/hazelcast/hazelcast-cpp-client/files/2533227/javaIMapErrOutput.txt)
[javaIMapPerformance.txt](https://github.com/hazelcast/hazelcast-cpp-client/files/2533228/javaIMapPerformance.txt)
[Client_java.txt](https://github.com/hazelcast/hazelcast-cpp-client/files/2533233/Client_java.txt)
[main_cpp.txt](https://github.com/hazelcast/hazelcast-cpp-client/files/2533234/main_cpp.txt)

**Comment by ihsandemir** [comment]
Hello @wsw2016 

I tried the example code that you provided in the following environments and my results are very different from yours. It is important that you use the Release version of the library and also build your executable as release to benefit from optimizations of the compiler. In order to do this, I downloaded the 3.10.1 release library from https://hazelcast.org/clients/cplusplus/ and I compiled and linked against this library (non-tls library) . I recommend that you do the same so that we have a fair comparison. When you manually build, there may be some build optimizations missing, hence it is always safe to use the released library. 

**My findings:** At Linux, you should do a run of multiple iterations and take average, otherwise your numbers are misleading. When I set number of iterations to for example 30, the numbers for get and set gets closer (There may be a slight difference for set since it does more serialization/deserialization than get operation). In performance tests, a single iteration is not a good idea but do the operations multiple times and taking an average is more healthy approach. Remember that the server cold start and many other factors may affect performance.

Can you please try this way and let me know how your results are?

The source code of the project code (Same as yours except that I added a command line parameter to pass the number of iterations):
[deneme.tar.gz](https://github.com/hazelcast/hazelcast-cpp-client/files/2571649/deneme.tar.gz)

**Results:**

Mac OS X (2.6 GHz Intel Core i5, 16 GB RAM): I have several other programs running during the test (such as some development IDEs, chrome, etc.)
**Mac OS X  Number of Iterations : 1**
```
/Users/ihsan/Desktop/work/src/deneme/cmake-build-release/deneme 1
Testing with number of iterations: 1
Nov 12, 2018 01:29:18 PM INFO: [140735732355904] hz.client_1[dev] [3.10.1] (20180828:f9f25cb2a2) LifecycleService::LifecycleEvent STARTING
Nov 12, 2018 01:29:18 PM INFO: [140735732355904] hz.client_1[dev] [3.10.1] LifecycleService::LifecycleEvent STARTED
Nov 12, 2018 01:29:18 PM INFO: [123145555234816] hz.client_1[dev] [3.10.1] Trying to connect to Address[localhost:5701] as owner member
Nov 12, 2018 01:29:18 PM INFO: [123145554698240] hz.client_1[dev] [3.10.1] Setting ClientConnection{alive=1, connectionId=1, remoteEndpoint=Address[192.168.1.119:5701], lastReadTime=2018-11-12 10:29:18.142000, closedTime=never, connected server version=3.12-SNAPSHOT} as owner with principal uuid: 2ce7df6d-5fb6-4e9b-852f-eff3aaf0ea71 ownerUuid: 46e9f77d-9763-48f7-9cf3-e3ad83381e1d
Nov 12, 2018 01:29:18 PM INFO: [123145554698240] hz.client_1[dev] [3.10.1] Authenticated with server Address[192.168.1.119:5701], server version:3.12-SNAPSHOT Local address: null
Nov 12, 2018 01:29:18 PM INFO: [123145557917696] hz.client_1[dev] [3.10.1] 

Members [1]  {
	Member[192.168.1.119]:5701 - 46e9f77d-9763-48f7-9cf3-e3ad83381e1d
}

Nov 12, 2018 01:29:18 PM INFO: [123145555234816] hz.client_1[dev] [3.10.1] LifecycleService::LifecycleEvent CLIENT_CONNECTED
1, 477, 0.00209644, 313, 0.00319489 MBytes/Sec
2, 280, 0.00714286, 213, 0.00938967 MBytes/Sec
4, 300, 0.0133333, 252, 0.015873 MBytes/Sec
8, 271, 0.0295203, 194, 0.0412371 MBytes/Sec
16, 232, 0.0689655, 195, 0.0820513 MBytes/Sec
32, 203, 0.157635, 197, 0.162437 MBytes/Sec
64, 211, 0.303318, 198, 0.323232 MBytes/Sec
128, 201, 0.636816, 192, 0.666667 MBytes/Sec
256, 205, 1.24878, 202, 1.26733 MBytes/Sec
512, 196, 2.61224, 203, 2.52217 MBytes/Sec
1024, 258, 3.96899, 214, 4.78505 MBytes/Sec
2048, 246, 8.3252, 234, 8.75214 MBytes/Sec
4096, 252, 16.254, 248, 16.5161 MBytes/Sec
8192, 285, 28.7439, 264, 31.0303 MBytes/Sec
16384, 378, 43.3439, 346, 47.3526 MBytes/Sec
32768, 612, 53.5425, 561, 58.41 MBytes/Sec
65536, 1044, 62.7739, 816, 80.3137 MBytes/Sec
131072, 2163, 60.5973, 1491, 87.9088 MBytes/Sec
262144, 3279, 79.9463, 2325, 112.75 MBytes/Sec
524288, 8024, 65.34, 6303, 83.1807 MBytes/Sec
1048576, 21036, 49.8467, 11755, 89.2026 MBytes/Sec
2097152, 23875, 87.8388, 18428, 113.802 MBytes/Sec
4194304, 48064, 87.265, 38912, 107.789 MBytes/Sec
8388608, 129295, 64.8796, 78749, 106.523 MBytes/Sec
16777216, 216638, 77.4436, 180179, 93.1142 MBytes/Sec

Nov 12, 2018 01:29:19 PM INFO: [140735732355904] hz.client_1[dev] [3.10.1] LifecycleService::LifecycleEvent SHUTTING_DOWN
Nov 12, 2018 01:29:19 PM INFO: [140735732355904] hz.client_1[dev] [3.10.1] Removed connection to endpoint: Address[192.168.1.119:5701], connection: ClientConnection{alive=0, connectionId=1, remoteEndpoint=Address[192.168.1.119:5701], lastReadTime=2018-11-12 10:29:18.979000, closedTime=2018-11-12 10:29:19.127000, connected server version=3.12-SNAPSHOT}
Nov 12, 2018 01:29:20 PM INFO: [140735732355904] hz.client_1[dev] [3.10.1] LifecycleService::LifecycleEvent SHUTDOWN

Process finished with exit code 0
```
**Mac OS X Number Of Iterations: 30**
```
/Users/ihsan/Desktop/work/src/deneme/cmake-build-release/deneme 30
Testing with number of iterations: 30
Nov 12, 2018 01:31:10 PM INFO: [140735732355904] hz.client_1[dev] [3.10.1] (20180828:f9f25cb2a2) LifecycleService::LifecycleEvent STARTING
Nov 12, 2018 01:31:10 PM INFO: [140735732355904] hz.client_1[dev] [3.10.1] LifecycleService::LifecycleEvent STARTED
Nov 12, 2018 01:31:10 PM INFO: [123145429807104] hz.client_1[dev] [3.10.1] Trying to connect to Address[localhost:5701] as owner member
Nov 12, 2018 01:31:10 PM INFO: [123145429270528] hz.client_1[dev] [3.10.1] Setting ClientConnection{alive=1, connectionId=1, remoteEndpoint=Address[192.168.1.119:5701], lastReadTime=2018-11-12 10:31:10.866000, closedTime=never, connected server version=3.12-SNAPSHOT} as owner with principal uuid: 781913d5-3ee4-4bdc-bd64-6787a4810e85 ownerUuid: 46e9f77d-9763-48f7-9cf3-e3ad83381e1d
Nov 12, 2018 01:31:10 PM INFO: [123145429270528] hz.client_1[dev] [3.10.1] Authenticated with server Address[192.168.1.119:5701], server version:3.12-SNAPSHOT Local address: null
Nov 12, 2018 01:31:10 PM INFO: [123145431953408] hz.client_1[dev] [3.10.1] 

Members [1]  {
	Member[192.168.1.119]:5701 - 46e9f77d-9763-48f7-9cf3-e3ad83381e1d
}

Nov 12, 2018 01:31:11 PM INFO: [123145429807104] hz.client_1[dev] [3.10.1] LifecycleService::LifecycleEvent CLIENT_CONNECTED
1, 6208, 0.00483247, 4304, 0.00697026 MBytes/Sec
2, 3829, 0.0156699, 3336, 0.0179856 MBytes/Sec
4, 3341, 0.0359174, 3363, 0.0356824 MBytes/Sec
8, 3840, 0.0625, 3204, 0.0749064 MBytes/Sec
16, 3449, 0.139171, 3403, 0.141052 MBytes/Sec
32, 3400, 0.282353, 3252, 0.295203 MBytes/Sec
64, 3452, 0.556199, 3240, 0.592593 MBytes/Sec
128, 3294, 1.16576, 3503, 1.0962 MBytes/Sec
256, 3519, 2.18244, 3444, 2.22997 MBytes/Sec
512, 3591, 4.27736, 3407, 4.50837 MBytes/Sec
1024, 3696, 8.31169, 3515, 8.73969 MBytes/Sec
2048, 3796, 16.1855, 3871, 15.8719 MBytes/Sec
4096, 4404, 27.9019, 4214, 29.1599 MBytes/Sec
8192, 5846, 42.039, 13265, 18.527 MBytes/Sec
16384, 9625, 51.067, 7794, 63.0639 MBytes/Sec
32768, 11859, 82.894, 11599, 84.7521 MBytes/Sec
65536, 23266, 84.5044, 24526, 80.1631 MBytes/Sec
131072, 40344, 97.4658, 43764, 89.8492 MBytes/Sec
262144, 92776, 84.7668, 78133, 100.653 MBytes/Sec
524288, 162212, 96.9635, 192671, 81.6347 MBytes/Sec
1048576, 317028, 99.2256, 284499, 110.571 MBytes/Sec
2097152, 581628, 108.17, 561146, 112.118 MBytes/Sec
4194304, 1371219, 91.7644, 1250978, 100.585 MBytes/Sec
8388608, 2663507, 94.4838, 2413564, 104.268 MBytes/Sec
16777216, 5453100, 92.2991, 4978946, 101.089 MBytes/Sec

Nov 12, 2018 01:31:31 PM INFO: [140735732355904] hz.client_1[dev] [3.10.1] LifecycleService::LifecycleEvent SHUTTING_DOWN
Nov 12, 2018 01:31:31 PM INFO: [140735732355904] hz.client_1[dev] [3.10.1] Removed connection to endpoint: Address[192.168.1.119:5701], connection: ClientConnection{alive=0, connectionId=1, remoteEndpoint=Address[192.168.1.119:5701], lastReadTime=2018-11-12 10:31:31.676000, closedTime=2018-11-12 10:31:31.816000, connected server version=3.12-SNAPSHOT}
Nov 12, 2018 01:31:33 PM INFO: [140735732355904] hz.client_1[dev] [3.10.1] LifecycleService::LifecycleEvent SHUTDOWN

Process finished with exit code 0
```

Furthermore, I tested using the docker container on the same MAc OS computer, docker running CentOS 7 image from: https://hub.docker.com/r/ihsan/centos7_gcc_485_ssl/

**CentOS 7 Number Of Iterations: 1**
```
[root@5a01c880395e build]# ./deneme 1
Testing with number of iterations: 1
Nov 12, 2018 11:03:36 AM INFO: [140325263963968] hz.client_1[dev] [3.10.1] (20180828:f9f25cb2) LifecycleService::LifecycleEvent STARTING
Nov 12, 2018 11:03:36 AM INFO: [140325263963968] hz.client_1[dev] [3.10.1] LifecycleService::LifecycleEvent STARTED
Nov 12, 2018 11:03:36 AM INFO: [140325214963456] hz.client_1[dev] [3.10.1] Trying to connect to Address[localhost:5701] as owner member
Nov 12, 2018 11:03:36 AM INFO: [140325240141568] hz.client_1[dev] [3.10.1] Setting ClientConnection{alive=1, connectionId=1, remoteEndpoint=Address[172.17.0.2:5701], lastReadTime=2018-11-12 11:03:37.000000, closedTime=never, connected server version=3.11} as owner with principal uuid: 6a4936f2-79b8-4743-8c66-6569a78f58b1 ownerUuid: b210c035-db65-46ae-9e73-ceb15f316627
Nov 12, 2018 11:03:36 AM INFO: [140325240141568] hz.client_1[dev] [3.10.1] Authenticated with server Address[172.17.0.2:5701], server version:3.11 Local address: Address[127.0.0.1:40118]
Nov 12, 2018 11:03:37 AM INFO: [140325181392640] hz.client_1[dev] [3.10.1] 

Members [1]  {
        Member[172.17.0.2]:5701 - b210c035-db65-46ae-9e73-ceb15f316627
}

Nov 12, 2018 11:03:37 AM INFO: [140325214963456] hz.client_1[dev] [3.10.1] LifecycleService::LifecycleEvent CLIENT_CONNECTED
1, 1398, 0.000715308, 903, 0.00110742 MBytes/Sec
2, 888, 0.00225225, 563, 0.0035524 MBytes/Sec
4, 2598, 0.00153965, 567, 0.00705467 MBytes/Sec
8, 646, 0.0123839, 809, 0.00988875 MBytes/Sec
16, 1386, 0.011544, 961, 0.0166493 MBytes/Sec
32, 1845, 0.0173442, 920, 0.0347826 MBytes/Sec
64, 933, 0.0685959, 689, 0.0928882 MBytes/Sec
128, 877, 0.145952, 649, 0.197227 MBytes/Sec
256, 780, 0.328205, 563, 0.454707 MBytes/Sec
512, 734, 0.697548, 607, 0.843493 MBytes/Sec
1024, 783, 1.30779, 606, 1.68977 MBytes/Sec
2048, 1081, 1.89454, 704, 2.90909 MBytes/Sec
4096, 3763, 1.08849, 875, 4.68114 MBytes/Sec
8192, 2345, 3.49339, 3040, 2.69474 MBytes/Sec
16384, 1659, 9.87583, 1028, 15.9377 MBytes/Sec
32768, 1536, 21.3333, 1617, 20.2647 MBytes/Sec
65536, 51147, 1.28133, 3035, 21.5934 MBytes/Sec
131072, 92651, 1.41469, 3756, 34.8967 MBytes/Sec
262144, 204911, 1.27931, 4906, 53.4333 MBytes/Sec
524288, 405916, 1.29162, 5710, 91.8193 MBytes/Sec
1048576, 804254, 1.30379, 12014, 87.2795 MBytes/Sec
2097152, 1615630, 1.29804, 21609, 97.0499 MBytes/Sec
4194304, 3241241, 1.29404, 119572, 35.0776 MBytes/Sec
8388608, 6438628, 1.30286, 73411, 114.269 MBytes/Sec
16777216, 12940478, 1.29649, 161393, 103.953 MBytes/Sec

Nov 12, 2018 11:04:03 AM INFO: [140325263963968] hz.client_1[dev] [3.10.1] LifecycleService::LifecycleEvent SHUTTING_DOWN
Nov 12, 2018 11:04:03 AM INFO: [140325263963968] hz.client_1[dev] [3.10.1] Removed connection to endpoint: Address[172.17.0.2:5701], connection: ClientConnection{alive=0, connectionId=1, remoteEndpoint=Address[172.17.0.2:5701], lastReadTime=2018-11-12 11:04:03.261000, closedTime=2018-11-12 11:04:03.384000, connected server version=3.11}
Nov 12, 2018 11:04:03 AM INFO: [140325263963968] hz.client_1[dev] [3.10.1] LifecycleService::LifecycleEvent SHUTDOWN
```

**CentOS 7 Number Of Iterations: 30**
```
[root@5a01c880395e build]# ./deneme 30
Testing with number of iterations: 30
Nov 12, 2018 11:04:45 AM INFO: [140526787610432] hz.client_1[dev] [3.10.1] (20180828:f9f25cb2) LifecycleService::LifecycleEvent STARTING
Nov 12, 2018 11:04:45 AM INFO: [140526787610432] hz.client_1[dev] [3.10.1] LifecycleService::LifecycleEvent STARTED
Nov 12, 2018 11:04:45 AM INFO: [140526738609920] hz.client_1[dev] [3.10.1] Trying to connect to Address[localhost:5701] as owner member
Nov 12, 2018 11:04:45 AM INFO: [140526763788032] hz.client_1[dev] [3.10.1] Setting ClientConnection{alive=1, connectionId=1, remoteEndpoint=Address[172.17.0.2:5701], lastReadTime=2018-11-12 11:04:45.427000, closedTime=never, connected server version=3.11} as owner with principal uuid: ff3442ac-c08b-4985-aba8-ad76369711d9 ownerUuid: b210c035-db65-46ae-9e73-ceb15f316627
Nov 12, 2018 11:04:45 AM INFO: [140526763788032] hz.client_1[dev] [3.10.1] Authenticated with server Address[172.17.0.2:5701], server version:3.11 Local address: Address[127.0.0.1:40124]
Nov 12, 2018 11:04:45 AM INFO: [140526730217216] hz.client_1[dev] [3.10.1] 

Members [1]  {
        Member[172.17.0.2]:5701 - b210c035-db65-46ae-9e73-ceb15f316627
}

Nov 12, 2018 11:04:45 AM INFO: [140526738609920] hz.client_1[dev] [3.10.1] LifecycleService::LifecycleEvent CLIENT_CONNECTED
1, 34686, 0.000864902, 29600, 0.00101351 MBytes/Sec
2, 38113, 0.00157427, 31597, 0.00189891 MBytes/Sec
4, 51906, 0.00231187, 22760, 0.00527241 MBytes/Sec
8, 23580, 0.0101781, 29901, 0.00802649 MBytes/Sec
16, 25227, 0.0190272, 29964, 0.0160192 MBytes/Sec
32, 33811, 0.0283931, 22672, 0.042343 MBytes/Sec
64, 26022, 0.0737837, 30858, 0.0622205 MBytes/Sec
128, 33427, 0.114877, 20238, 0.189742 MBytes/Sec
256, 23807, 0.322594, 41507, 0.185029 MBytes/Sec
512, 18293, 0.839665, 20742, 0.740526 MBytes/Sec
1024, 35784, 0.858484, 18449, 1.66513 MBytes/Sec
2048, 42991, 1.42914, 20896, 2.94028 MBytes/Sec
4096, 22737, 5.40441, 26038, 4.71926 MBytes/Sec
8192, 27819, 8.83425, 26109, 9.41285 MBytes/Sec
16384, 33817, 14.5347, 32839, 14.9676 MBytes/Sec
32768, 41883, 23.4711, 35441, 27.7374 MBytes/Sec
65536, 53018, 37.0833, 40666, 48.347 MBytes/Sec
131072, 108755, 36.1561, 47883, 82.1202 MBytes/Sec
262144, 120469, 65.2809, 85576, 91.8987 MBytes/Sec
524288, 181866, 86.4848, 153222, 102.653 MBytes/Sec
1048576, 373864, 84.141, 289605, 108.621 MBytes/Sec
2097152, 651979, 96.4978, 553821, 113.601 MBytes/Sec
4194304, 1428373, 88.0926, 1168564, 107.678 MBytes/Sec
8388608, 2470661, 101.859, 2208790, 113.935 MBytes/Sec
16777216, 5499379, 91.5224, 4503709, 111.756 MBytes/Sec

Nov 12, 2018 11:05:06 AM INFO: [140526787610432] hz.client_1[dev] [3.10.1] LifecycleService::LifecycleEvent SHUTTING_DOWN
Nov 12, 2018 11:05:06 AM INFO: [140526787610432] hz.client_1[dev] [3.10.1] Removed connection to endpoint: Address[172.17.0.2:5701], connection: ClientConnection{alive=0, connectionId=1, remoteEndpoint=Address[172.17.0.2:5701], lastReadTime=2018-11-12 11:05:06.341000, closedTime=2018-11-12 11:05:06.448000, connected server version=3.11}
Nov 12, 2018 11:05:07 AM INFO: [140526787610432] hz.client_1[dev] [3.10.1] LifecycleService::LifecycleEvent SHUTDOWN
```

**Comment by ihsandemir** [comment]
@wsw2016 For getting the most performance we need to write a test that does multi thread set and get so that we can utilize the client fully. Hence, you may get closer to wire utilization this way.

**Comment by wsw2016** [comment]
Make an average to remove random deviations is ok, but if you start the test with 1 iteration 30 times you will get 13 seconds to transfer the last 16 MBytes message in all cases. In my setup I don't get this speed improvement with 30 iterations.

I think I found the bottleneck - the combination of non-blocking TCP-Socket with _select_ seems to be broken for small sending buffers. In TcpSocket.cpp you set the buffer to 32kB, Linux itself uses 1MB. Commenting out the lines
if (::setsockopt(socketId, SOL_SOCKET, SO_SND_BUF, (char *) &size, sizeof(size)) {
  ....
}
solves the problem at least partially. I get 14 MBytes/sec instead of 1.4. I measured the timings, with small buffer the socket sends only ~64 kB at once and each following _select_ takes 40 milliseconds. I don't know, how important is for the CppClient to set the SendBuffer instead of keeping the default value. Perhaps you remove the line, at least for Linux.

**Comment by wsw2016** [comment]
It goes event upto 64 MBytes/sec in release mode - the last line for 16MB message is on i5 Ubuntu in VirtualBox:
16777216, 263157, **63.7536**, 174736, **96.0147** MBytes/Sec

**Comment by ihsandemir** [comment]
Hello @wsw2016 

Thanks for the feedback and pointing out the cause of the problem. I got a chance to try on my local machine with different socket read/write buffer sizes. Larger sizes makes the performance for the large byte buffer tests better performing as you pointed out. I need to perform more study on performance improvements.

In the meanwhile, I also examined how our Java client uses the buffer size. I found that Java uses 128 K buffer size which performs better and it also provides the https://github.com/hazelcast/hazelcast/blob/master/hazelcast-client/src/main/java/com/hazelcast/client/config/SocketOptions.java#L141 and the SocketOptions class which allows the user be able to configure the socket options. Hence, as a solution I will put this class in place with the default buffer size being 128 K as in Java client. I will submit a PR for this and it should fix the problem that you found.
